File: github-push.py
#!/usr/bin/env python3
"""
github-push.py - 通过 GitHub API 推送代码 (解决 git HTTPS 连接问题)
用法: python3 github-push.py [提交信息]
"""
import os, sys, json, urllib.request, base64, subprocess, re

def api(url, data=None, method="GET"):
    token = os.environ.get("GH_TOKEN")
    if not token:
        # 从 git remote URL 提取
        try:
            remote = subprocess.run(["git", "remote", "get-url", "origin"],
                                  capture_output=True, text=True, cwd=REPO_DIR).stdout.strip()
            m = re.search(r'https://[^:]+:([^@]+)@github\.com/([^/]+)/([^/]+?)\.git', remote)
            if m:
                token = m.group(1)
                globals()["OWNER"] = m.group(2)
                globals()["REPO"] = m.group(3)
        except: pass
    if not token:
        print("❌ 请设置 GH_TOKEN 环境变量或在 git remote URL 中包含 token")
        sys.exit(1)

    headers = {
        "Authorization": f"Bearer {token}",
        "Accept": "application/vnd.github+json",
        "User-Agent": "github-push-script",
    }
    if data is not None:
        headers["Content-Type"] = "application/json"
        req = urllib.request.Request(f"https://api.github.com{url}", json.dumps(data).encode(), headers, method=method)
    else:
        req = urllib.request.Request(f"https://api.github.com{url}", headers=headers, method=method)
    try:
        with urllib.request.urlopen(req) as r:
            return json.loads(r.read().decode()) if r.readable() else {}
    except urllib.error.HTTPError as e:
        err = e.read().decode()
        print(f"⚠️ API 错误 ({e.code}): {e.url}")
        # 404 时返回 None 表示不存在
        if e.code == 404: return None
        print(err[:200])
        return None
    except Exception as e:
        print(f"⚠️ 请求错误: {e}")
        return None

REPO_DIR = os.getcwd()
OWNER, REPO = "", ""
COMMIT_MSG = sys.argv[1] if len(sys.argv) > 1 else "更新代码"

print(f"📂 仓库目录: {REPO_DIR}")
print(f"📝 提交信息: {COMMIT_MSG}")

# 获取文件列表
result = subprocess.run(["git", "ls-files"], capture_output=True, text=True, cwd=REPO_DIR)
files = [f for f in result.stdout.strip().split("\n") if f and not f.startswith(".git/")]
if not files:
    # fallback: 所有非 git 文件
    for root, dirs, fnames in os.walk(REPO_DIR):
        dirs[:] = [d for d in dirs if d != ".git"]
        for fn in fnames:
            fp = os.path.relpath(os.path.join(root, fn), REPO_DIR)
            if fp != "github-push.py" and fp != "git-api-push.sh":
                files.append(fp)

print(f"📄 共 {len(files)} 个文件")
for f in files:
    print(f"   - {f}")

# 获取当前分支信息
print("🔍 获取仓库信息...")
ref = api(f"/repos/{OWNER}/{REPO}/git/refs/heads/master")
if ref is None:
    print("🆕 空仓库，创建初始提交")

base_tree_sha = None
parent_sha = None
if ref and "object" in ref:
    parent_sha = ref["object"]["sha"]
    commit = api(f"/repos/{OWNER}/{REPO}/git/commits/{parent_sha}")
    if commit and "tree" in commit:
        base_tree_sha = commit["tree"]["sha"]

# 为每个文件创建 blob
tree_entries = []
for f in files:
    fpath = os.path.join(REPO_DIR, f)
    if not os.path.isfile(fpath):
        continue
    with open(fpath, "rb") as fh:
        content = fh.read()
    b64 = base64.b64encode(content).decode()

    print(f"  ⏳ 上传: {f}")
    blob = api(f"/repos/{OWNER}/{REPO}/git/blobs", {"content": b64, "encoding": "base64"}, "POST")
    if blob and "sha" in blob:
        tree_entries.append({"path": f, "mode": "100644", "type": "blob", "sha": blob["sha"]})
    else:
        print(f"  ⚠️  跳过 {f} (blob 创建失败)")

if not tree_entries:
    print("❌ 没有文件成功上传")
    sys.exit(1)

# 创建新树
print("🌳 创建树对象...")
tree_data = {"tree": tree_entries}
if base_tree_sha:
    tree_data["base_tree"] = base_tree_sha
new_tree = api(f"/repos/{OWNER}/{REPO}/git/trees", tree_data, "POST")
if not new_tree or "sha" not in new_tree:
    print("❌ 创建树失败")
    sys.exit(1)
print(f"  ✅ 树 SHA: {new_tree['sha']}")

# 创建 commit
print("💾 创建提交...")
commit_data = {
    "message": COMMIT_MSG,
    "tree": new_tree["sha"],
    "parents": [parent_sha] if parent_sha else [],
}
new_commit = api(f"/repos/{OWNER}/{REPO}/git/commits", commit_data, "POST")
if not new_commit or "sha" not in new_commit:
    print("❌ 创建提交失败")
    sys.exit(1)
print(f"  ✅ Commit SHA: {new_commit['sha']}")

# 更新分支
print("📤 推送到 master...")
result = api(f"/repos/{OWNER}/{REPO}/git/refs/heads/master",
             {"sha": new_commit["sha"], "force": True}, "PATCH")
if result and "object" in result and result["object"]["sha"] == new_commit["sha"]:
    print()
    print("✅✅✅ 推送成功！ ✅✅✅")
    print(f"   https://github.com/{OWNER}/{REPO}")
    # 更新本地 git 引用
    subprocess.run(["git", "update-ref", "refs/heads/master", new_commit["sha"]],
                   capture_output=True, cwd=REPO_DIR)
else:
    print("⚠️ 推送可能失败:")
    print(json.dumps(result, indent=2)[:200])
    sys.exit(1)

File: github-push.js
#!/usr/bin/env node
/**
 * github-push.js - 通过 GitHub API 推送代码
 * 用法: node github-push.js [提交信息]
 *
 * 环境变量: GH_TOKEN 或 git remote URL 中包含 token
 */
const fs = require('fs');
const path = require('path');
const { execSync } = require('child_process');

const REPO_DIR = process.cwd();
const COMMIT_MSG = process.argv[2] || '更新代码';

// 获取 Token
function getTokenAndRepo() {
  let token = process.env.GH_TOKEN || '';
  let owner = '', repo = '';

  try {
    const remote = execSync('git remote get-url origin', { cwd: REPO_DIR }).toString().trim();
    const match = remote.match(/https:\/\/([^:]+):([^@]+)@github\.com\/([^/]+)\/([^/]+?)\.git/);
    if (match) {
      token = token || match[2];
      owner = match[3];
      repo = match[4];
    } else {
      const match2 = remote.match(/github\.com\/([^/]+)\/([^/]+?)\.git/);
      if (match2) {
        owner = match2[1];
        repo = match2[2];
      }
    }
  } catch (e) { /* no remote */ }

  if (!token && !process.env.GH_TOKEN) {
    console.error('❌ 请设置 GH_TOKEN 环境变量或 git remote URL 中包含 token');
    console.error('   export GH_TOKEN=ghp_xxx');
    process.exit(1);
  }

  return { token: token || process.env.GH_TOKEN, owner, repo };
}

// GitHub API 调用
async function api(path, data, method = 'GET') {
  const url = `https://api.github.com${path}`;
  const headers = {
    Authorization: `Bearer ${CONFIG.token}`,
    Accept: 'application/vnd.github+json',
    'User-Agent': 'github-push-js',
    'Content-Type': 'application/json',
  };

  const opts = { method, headers };
  if (data !== undefined) opts.body = JSON.stringify(data);

  const resp = await fetch(url, opts);
  if (!resp.ok && resp.status === 404) return null;
  if (!resp.ok) {
    const text = await resp.text();
    console.error(`⚠️ API 错误 ${resp.status}: ${path}`);
    console.error(text.slice(0, 200));
    return null;
  }
  return resp.json();
}

const CONFIG = getTokenAndRepo();
const { owner, repo } = CONFIG;

// 主函数
async function main() {
  console.log(`📂 目录: ${REPO_DIR}`);
  console.log(`📝 提交: ${COMMIT_MSG}`);
  console.log(`📦 仓库: ${owner}/${repo}\n`);

  // 获取文件列表
  let files = [];
  try {
    const out = execSync('git ls-files', { cwd: REPO_DIR }).toString();
    files = out.split('\n').filter(f => f && !f.startsWith('.git/'));
  } catch (e) { /* not a git repo or no files tracked */ }

  if (files.length === 0) {
    // fallback: scan directory
    files = fs.readdirSync(REPO_DIR)
      .filter(f => !f.startsWith('.git') && f !== 'github-push.js' && f !== 'git-api-push.sh')
      .filter(f => fs.statSync(path.join(REPO_DIR, f)).isFile());
  }

  console.log(`📄 共 ${files.length} 个文件`);
  files.forEach(f => console.log(`   - ${f}`));

  // 获取当前分支信息
  console.log('\n🔍 获取仓库信息...');
  const ref = await api(`/repos/${owner}/${repo}/git/refs/heads/master`);
  let parentSha = null;
  let baseTreeSha = null;

  if (ref && ref.object) {
    parentSha = ref.object.sha;
    const commit = await api(`/repos/${owner}/${repo}/git/commits/${parentSha}`);
    if (commit && commit.tree) baseTreeSha = commit.tree.sha;
    console.log(`   父提交: ${parentSha.slice(0, 8)}`);
  } else {
    console.log('   🆕 首次提交');
  }

  // 为每个文件创建 blob
  const treeEntries = [];
  for (const f of files) {
    const fpath = path.join(REPO_DIR, f);
    if (!fs.existsSync(fpath)) continue;

    const content = fs.readFileSync(fpath);
    const b64 = content.toString('base64');

    process.stdout.write(`   ⏳ 上传: ${f}`);
    const blob = await api(`/repos/${owner}/${repo}/git/blobs`, { content: b64, encoding: 'base64' }, 'POST');
    if (blob && blob.sha) {
      treeEntries.push({ path: f, mode: '100644', type: 'blob', sha: blob.sha });
      process.stdout.write(' ✅\n');
    } else {
      process.stdout.write(' ⚠️ 失败\n');
    }
  }

  if (treeEntries.length === 0) {
    console.error('\n❌ 没有文件成功上传');
    process.exit(1);
  }

  // 创建新树
  console.log('\n🌳 创建树对象...');
  const treeData = { tree: treeEntries };
  if (baseTreeSha) treeData.base_tree = baseTreeSha;
  const newTree = await api(`/repos/${owner}/${repo}/git/trees`, treeData, 'POST');
  if (!newTree || !newTree.sha) {
    console.error('❌ 创建树失败');
    process.exit(1);
  }
  console.log(`   ✅ 树 SHA: ${newTree.sha}`);

  // 创建 commit
  console.log('💾 创建提交...');
  const commitData = { message: COMMIT_MSG, tree: newTree.sha, parents: parentSha ? [parentSha] : [] };
  const newCommit = await api(`/repos/${owner}/${repo}/git/commits`, commitData, 'POST');
  if (!newCommit || !newCommit.sha) {
    console.error('❌ 创建提交失败');
    process.exit(1);
  }
  console.log(`   ✅ Commit SHA: ${newCommit.sha}`);

  // 更新分支
  console.log('📤 推送到 master...');
  const result = await api(`/repos/${owner}/${repo}/git/refs/heads/master`,
    { sha: newCommit.sha, force: true }, 'PATCH');

  if (result && result.object && result.object.sha === newCommit.sha) {
    console.log('\n✅✅✅ 推送成功！ ✅✅✅');
    console.log(`   https://github.com/${owner}/${repo}\n`);
    // 更新本地 git 引用
    try {
      execSync(`git update-ref refs/heads/master ${newCommit.sha}`, { cwd: REPO_DIR });
    } catch (e) { /* ignore */ }
  } else {
    console.error('\n⚠️ 推送可能失败');
    if (result) console.error(JSON.stringify(result, null, 2).slice(0, 200));
    process.exit(1);
  }
}

main().catch(e => {
  console.error('❌ 错误:', e.message);
  process.exit(1);
});
